jedi/refactoring.py

```diff
@@ -12,86 +12,146 @@
+
+
+class FileState(namedtuple('FileState', ['path', 'lines'])):
+    @property
+    def content_str(self):
+        return '\n'.join(self.lines)
+
+    def __eq__(self, other):
+        if not isinstance(other, FileState):
+            return False
+        return self.path == other.path and \
+               len(self.lines) == len(other.lines) and \
+               all((l1 == l2 for l1, l2 in zip(self.lines, other.lines)))
+
+
+class Change(namedtuple('Change', ['old_state', 'new_state'])):
+    @property
+    def diff(self):
+        return unified_diff(self.old_state.lines, self.new_state.lines)
+
+    def __eq__(self, other):
+        if not isinstance(other, Change):
+            return False
+        return self.old_state == other.old_state and \
+               self.new_state == other.new_state
+
+
+class Changes(object):
+    def __init__(self, changes):
+        self.changes = changes
 
     def new_files(self):
-        dct = {}
-        for old_path, (new_path, old_l, new_l) in self.change_dct.items():
-            dct[new_path] = '\n'.join(new_l)
-        return dct
+        return {ch.new_state.path: ch.new_state.content_str
+                for ch in self.changes}
+
+    def old_files(self):
+        return {ch.old_state.path: ch.old_state.content_str
+                for ch in self.changes}
 
     def diff(self):
-        texts = []
-        for old_path, (new_path, old_l, new_l) in self.change_dct.items():
-            if old_path:
-                udiff = difflib.unified_diff(old_l, new_l)
-            else:
-                udiff = difflib.unified_diff(old_l, new_l, old_path, new_path)
-            texts.append('\n'.join(udiff))
-        return '\n'.join(texts)
+        return '\n'.join([ch.content_str for ch in self.changes])
 
 
-def rename(script, new_name):
-    """ The `args` / `kwargs` params are the same as in `api.Script`.
-    :param operation: The refactoring operation to execute.
-    :type operation: str
-    :type source: str
-    :return: list of changed lines/changed files
-    """
-    return Refactoring(_rename(script.usages(), new_name))
+def changes_for(refactoring, *args, **kwargs):
+    return Changes(refactoring(*args, **kwargs))
 
 
-def _rename(names, replace_str):
-    """ For both rename and inline. """
-    order = sorted(names, key=lambda x: (x.module_path, x.line, x.column),
-                   reverse=True)
+def rename(script, new_name):
+    def by_module_path(script):
+        return script.module_path
 
-    def process(path, old_lines, new_lines):
-        if new_lines is not None:  # goto next file, save last
-            dct[path] = path, old_lines, new_lines
+    usages = (u for u in sorted(script.usages(), key=by_module_path)
+              if not u.in_builtin_module())
+    usages_by_file = groupby(usages, by_module_path)
+    return [change_for_rename(m_path, usages, new_name)
+            for m_path, usages in usages_by_file]
```

Comment:
Hmm I think you haven't understood how the parser works, yet. What you should do is use the `usages` to modify the AST and then call `get_code()` on the AST. This is way simpler than what you did. Clearly this is also possible, but you make your life very complicated. Basically you don't need a Content and FileState class. It's way simpler. Do you understand what I mean?

Maybe just start with something simple: Parse something, change a Leaf.prefix and call `Parser.module.get_code()` and you will maybe see what I mean.